### PR TITLE
Create FileBaton to synchronize distributed JIT C++ extension builds

### LINKED ISSUE #6353: [jit cpp extensions] Possible file corruption in distributed setting
I believe there might be some sort of race condition that might happen when compiling jit cpp extensions in a distributed setting.

Here a summary of what happens:
- I launch my code from multiple different processes (to use with DistributedDataParallel, via `torch.distributed.launch`)
- each process starts to compile the extension
- they all try to save the extension to the same file
- sometimes, the file gets corrupted because of that

Here is part of the stack trace of the error I obtain (the rest is just the same error repeated):
```
  1 0: Traceback (most recent call last):
  2 0:   File "/private/home/fmassa/github/detectron.pytorch/torch_detectron/training_faster_rcnn.py", line 190, in <module>
  3 0:     from lib.faster_rcnn import fasterrcnn_resnet18, C2ResNetFasterRCNN
  4 0:   File "/private/home/fmassa/github/detectron.pytorch/torch_detectron/lib/faster_rcnn.py", line 13, in <module>
  5 0:     from lib.layers import ROIAlign, ROIPool, FixedBatchNorm2d
  6 0:   File "/private/home/fmassa/github/detectron.pytorch/torch_detectron/lib/layers/__init__.py", line 28, in <module>
  7 0:     _C = _load_C_extensions()
  8 0:   File "/private/home/fmassa/github/detectron.pytorch/torch_detectron/lib/layers/__init__.py", line 26, in _load_C_extensions
  9 0:     '-gencode arch=compute_70,code=sm_70'])
 10 0:   File "/private/home/fmassa/.conda/envs/detectron_v2/lib/python3.6/site-packages/torch/utils/cpp_extension.py", line 330, in load
 11 0:     build_directory = _get_build_directory(name, verbose)
 12 0:   File "/private/home/fmassa/.conda/envs/detectron_v2/lib/python3.6/site-packages/torch/utils/cpp_extension.py", line 392, in _get_build_directory
 13 0:     os.makedirs(build_directory)
 14 0:   File "/private/home/fmassa/.conda/envs/detectron_v2/lib/python3.6/os.py", line 220, in makedirs
 15 0:     mkdir(name, mode)
 16 0: FileExistsError: [Errno 17] File exists: '/tmp/torch_extensions/detectron_modules'
 17 0: Traceback (most recent call last):
 18 0:   File "/private/home/fmassa/github/detectron.pytorch/torch_detectron/training_faster_rcnn.py", line 190, in <module>
 19 0:     from lib.faster_rcnn import fasterrcnn_resnet18, C2ResNetFasterRCNN
 20 0:   File "/private/home/fmassa/github/detectron.pytorch/torch_detectron/lib/faster_rcnn.py", line 13, in <module>
 21 0:     from lib.layers import ROIAlign, ROIPool, FixedBatchNorm2d
 22 0:   File "/private/home/fmassa/github/detectron.pytorch/torch_detectron/lib/layers/__init__.py", line 28, in <module>
 23 0:     _C = _load_C_extensions()
 24 0:   File "/private/home/fmassa/github/detectron.pytorch/torch_detectron/lib/layers/__init__.py", line 26, in _load_C_extensions
 25 0:     '-gencode arch=compute_70,code=sm_70'])
 26 0:   File "/private/home/fmassa/.conda/envs/detectron_v2/lib/python3.6/site-packages/torch/utils/cpp_extension.py", line 361, in load
 27 0:     return _import_module_from_library(name, build_directory)
 28 0:   File "/private/home/fmassa/.conda/envs/detectron_v2/lib/python3.6/site-packages/torch/utils/cpp_extension.py", line 414, in _import_module_from_library
 29 0:     return imp.load_module(module_name, file, path, description)
 30 0:   File "/private/home/fmassa/.conda/envs/detectron_v2/lib/python3.6/imp.py", line 243, in load_module
 31 0:     return load_dynamic(name, filename, file)
 32 0:   File "/private/home/fmassa/.conda/envs/detectron_v2/lib/python3.6/imp.py", line 343, in load_dynamic
 33 0:     return _load(spec)
 34 0: ImportError: /tmp/torch_extensions/detectron_modules/detectron_modules.so: file too short
...
```
We can see that two types of errors pop out: `FileExistsError` and `ImportError: modules.so: file too short`.

Let me know if you need a minimal reproducible example.

pytorch version `'0.4.0a0+b21e135'`

### PR DESCRIPTION
**Pull request by goldsborough**
As reported by @fmassa in #6353 , when using JIT C++ extensions together with `torch.distributed.launch`, each process separately compiles the extension, and then crashes occasionally as the processes race with each other.

This PR creates a `FileBaton` datastructure, which allows two things:
1) atomically creating a file,
2) waiting until the file is deleted (spinning with sleeps)

This serves as a sufficient "rendezvous"/single-shot synchronization mechanism. As far as I can google, the file creation mechanism I use `O_CREAT | O_EXCL` is only atomic on Unix, so while this code will work fine on Windows, it may not behave in an entirely race-free way. I did not see a straightforward way provided by Python to do this on Windows (cc @peterjc123 )

Fixes #6353 

## Test plan

1. See #6353 for a MWE, also pass `verbose=True` to `load_ext()`
2. Run with `python -m torch.distributed.launch --nproc_per_node=4 mwe.py` and old code,
3. Observe multiple, concurrent compilation. Not necessarily a crash, but possibly,
4. Run with new code,
5. Observe that only one process compiles the code.

### CODE REVIEW

**Comment by peterjc123** [comment]
@goldsborough Well, i guess it's fine at the current state, because Windows distributed is not supported. And for the atomic file support, we will need some windows specfic flags. You can take a look [here](https://github.com/pytorch/pytorch/pull/3447/files#diff-f2a659b204c435cc38f1900702458dc3R1).

**Comment by peterjc123** [comment]
@goldsborough [Here](https://github.com/WoLpH/portalocker/blob/develop/portalocker/utils.py#L23) is an example of a cross-platform atomic `open` function.

**Comment by goldsborough** [comment]
Cool, thanks, I'll use that

**Comment by goldsborough** [comment]
Actually, no. It's important to note that we don't need atomic access (read or write) to a file, we need to atomically *create* a file, i.e. when two processes concurrently try to create the same file with the same name, only one must succeed and the other must fail. The code you linked is for atomically reading/writing to a file. It uses `NamedTemporaryFile`, which uses `mkstemp`, which relies on the `O_EXCL` functionality. I am also using this same flag. It is available on Windows and Unix, I'm just not sure it does the same thing on Windows as on Unix. Though I assume if it's available on Windows, it also does the same thing.

File: torch/utils/file_baton.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+    def wait(self):
+        '''
+        Periodically sleeps for a certain amount until the baton is released.
+
+        The amount of time slept depends on the ``wait_seconds`` parameter
+        passed to the constructor.
+        '''
+        while os.path.exists(self.lock_file_path):
+            time.sleep(self.wait_seconds)
```
Comment (apaszke) [inline torch/utils/file_baton.py]: That's not very elegant, but oh well 😄

File: torch/utils/cpp_extension.py [hunk 2]
```diff
@@ -168,16 +177,22 @@ def spawn(cmd):
                 # Using regex to match src, obj and include files
 
                 src_regex = re.compile('/T(p|c)(.*)')
-                src_list = [m.group(2) for m in (
-                    src_regex.match(elem) for elem in cmd) if m]
+                src_list = [
+                    m.group(2) for m in (src_regex.match(elem) for elem in cmd)
+                    if m
```
Comment (apaszke) [inline torch/utils/cpp_extension.py]: lol really? we accept lines up to 120 chars, and that really looks silly

**Review by apaszke — APPROVED** [review]
Please please please please when you use an auto-formatting tool, then make sure it doesn't do any dumb stuff. It's clear that this one doesn't respect our configuration, because it changed code that looks fine and passes our lint.

File: torch/utils/cpp_extension.py (hunk 2 above)
Comment (goldsborough) [reply torch/utils/cpp_extension.py]: Hmm I let Nuclide do this, I need to check if it uses our formatting guidelines

**Review by fmassa — APPROVED** [review]
I think this looks good, and is similar in spirit to what I hacked around to fix my problem. Thanks!

**Pull request merged**